pkg/volume/rbd/rbd_util.go

```diff
@@ -233,13 +352,31 @@ func (util *RBDUtil) AttachDisk(b rbdMounter) (string, error) {
 		}
 	}
 
-	devicePath, found := waitForPath(b.Pool, b.Image, 1)
-	if !found {
-		_, err = b.exec.Run("modprobe", "rbd")
-		if err != nil {
-			glog.Warningf("rbd: failed to load rbd kernel module:%v", err)
-		}
+	// If rbd-nbd tools are installed, we will try to map as a nbd device first.
+	nbdToolsFound := false
+	_, err = b.exec.Run("modprobe", "nbd")
+	if err != nil {
+		glog.V(5).Infof("rbd-nbd: nbd modprobe failed with error %v", err)
+	} else if _, err := b.exec.Run("rbd-nbd", "--version"); err != nil {
+		glog.V(5).Infof("rbd-nbd: getting rbd-nbd version failed with error %v", err)
+	} else {
+		glog.V(3).Infof("rbd-nbd tools were found. Attempting to map using rbd-nbd")
+		nbdToolsFound = true
```

Comment:
This is changing the default mapper to use the rbd-nbd by default. Seems pretty risky to me since the kernel RBD has been the default and only mapper from day 1.

What if an user want to use kernel RBD for mapping? The only way that could happen would be to uninstall rbd-nbd from the machines.

I would propose to introduce a flag? `useNBD=true` or something similar. This can be specified as a plugin parameter. The default would be `false` so that we keep using the kernel rbd.

Reply:
We've pulled some inspiration from https://github.com/kubernetes/kubernetes/pull/55866, to make it seamless to shift to user-mode driver when the appropriate host utils are installed, indicating cluster manager intent.

We also considered introducing a flag. This change would require updates to the k8s API, as well as configuration complexity.